[web-animations] allow an AcceleratedEffect to be cloned

graouts opened the pull request:
#### 97485effe50c457b74329522515d89b8be48bb3b
<pre>
[web-animations] allow an AcceleratedEffect to be cloned
<a href="https://bugs.webkit.org/show_bug.cgi?id=253567">https://bugs.webkit.org/show_bug.cgi?id=253567</a>

Reviewed by Dean Jackson.

When any AcceleratedEffect object will be decoded as part of the remote layer tree transaction,
we will need to transfer ownership to the thread used to resolve animations. We need to clone
any ref-counted object that would otherwise be marked as having been created on the main thread.

* Source/WebCore/platform/animation/AcceleratedEffect.cpp:
(WebCore::AcceleratedEffectKeyframe::clone const):
(WebCore::AcceleratedEffect::clone const):
(WebCore::AcceleratedEffect::copyWithProperties const):
(WebCore::AcceleratedEffect::copyWithProperties): Deleted.
* Source/WebCore/platform/animation/AcceleratedEffect.h:
* Source/WebCore/platform/animation/AcceleratedEffectValues.cpp:
(WebCore::AcceleratedEffectValues::clone const):
(WebCore::nonCalculatedLengthPoint): Deleted.
* Source/WebCore/platform/animation/AcceleratedEffectValues.h:
* Source/WebCore/platform/graphics/filters/FilterOperations.cpp:
(WebCore::FilterOperations::FilterOperations):
* Source/WebCore/platform/graphics/filters/FilterOperations.h:
(WebCore::FilterOperations::FilterOperations):
* Source/WebCore/rendering/PathOperation.cpp:
(WebCore::ReferencePathOperation::clone const):
(WebCore::RayPathOperation::clone const):
* Source/WebCore/rendering/PathOperation.h:
* Source/WebCore/rendering/style/BasicShapes.cpp:
(WebCore::BasicShapeCircle::clone const):
(WebCore::BasicShapeEllipse::clone const):
(WebCore::BasicShapePolygon::clone const):
(WebCore::BasicShapePath::clone const):
(WebCore::BasicShapeInset::clone const):
* Source/WebCore/rendering/style/BasicShapes.h:

Canonical link: <a href="https://commits.webkit.org/261418@main">https://commits.webkit.org/261418@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6974466dc2a766a453716af964d4bf496e7c0dda

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/111673 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/20809 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/90/builds/285 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/3428 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/120409 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/115734 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/22164 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/11889 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/86/builds/3165 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/117439 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/16459 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/99608 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/104582 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/98411 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/88/builds/187 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/45355 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/13288 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/89/builds/184 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/87789 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/13778 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/9629 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/19223 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/52182 "Passed tests") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/15768 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/4333 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

grorg commented on Source/WebCore/platform/animation/AcceleratedEffect.cpp:
```diff
@@ -96,7 +111,27 @@ Ref<AcceleratedEffect> AcceleratedEffect::create(Vector<AcceleratedEffectKeyfram
+Ref<AcceleratedEffect> AcceleratedEffect::clone() const
+{
+    Vector<AcceleratedEffectKeyframe> clonedKeyframes;
+    clonedKeyframes.reserveCapacity(m_keyframes.size());
```
Can't you do this in one line?

graouts replied:
Yes, using map(). Fixed in newer commit.

grorg commented on Source/WebCore/platform/animation/AcceleratedEffect.cpp:
```diff
@@ -96,7 +111,27 @@ Ref<AcceleratedEffect> AcceleratedEffect::create(Vector<AcceleratedEffectKeyfram
+    Vector<AcceleratedEffectKeyframe> clonedKeyframes;
+    clonedKeyframes.reserveCapacity(m_keyframes.size());
+    for (auto& keyframe : m_keyframes)
+        clonedKeyframes.uncheckedAppend(keyframe.clone());
```
Or use `map`?

graouts replied:
Good idea!

grorg commented on Source/WebCore/rendering/PathOperation.cpp:
```diff
@@ -45,6 +45,15 @@ Ref<ReferencePathOperation> ReferencePathOperation::create(std::optional<Path>&&
+{
+    if (auto path = this->path()) {
+        auto pathCopy = *path;
+        return adoptRef(*new ReferencePathOperation(WTFMove(pathCopy)));
```
Why not do this in one line?

grorg commented on Source/WebCore/rendering/PathOperation.cpp:
```diff
@@ -71,6 +80,13 @@ Ref<RayPathOperation> RayPathOperation::create(float angle, Size size, bool isCo
+{
+    auto containingBlockBoundingRect = m_containingBlockBoundingRect;
+    auto position = m_position;
+    return adoptRef(*new RayPathOperation(m_angle, m_size, m_isContaining, WTFMove(containingBlockBoundingRect), WTFMove(position)));
```
Do you gain anything with the local variables and the move here? Is it just because the method takes `&&`?

graouts replied:
Yes, I need copies to be made before they are moved.

graouts committed:
[web-animations] allow an AcceleratedEffect to be cloned
https://bugs.webkit.org/show_bug.cgi?id=253567

Reviewed by Dean Jackson.

When any AcceleratedEffect object will be decoded as part of the remote layer tree transaction,
we will need to transfer ownership to the thread used to resolve animations. We need to clone
any ref-counted object that would otherwise be marked as having been created on the main thread.

* Source/WebCore/platform/animation/AcceleratedEffect.cpp:
(WebCore::AcceleratedEffectKeyframe::clone const):
(WebCore::AcceleratedEffect::clone const):
(WebCore::AcceleratedEffect::copyWithProperties const):
(WebCore::AcceleratedEffect::copyWithProperties): Deleted.
* Source/WebCore/platform/animation/AcceleratedEffect.h:
* Source/WebCore/platform/animation/AcceleratedEffectValues.cpp:
(WebCore::AcceleratedEffectValues::clone const):
(WebCore::nonCalculatedLengthPoint): Deleted.
* Source/WebCore/platform/animation/AcceleratedEffectValues.h:
* Source/WebCore/platform/graphics/filters/FilterOperations.cpp:
(WebCore::FilterOperations::FilterOperations):
* Source/WebCore/platform/graphics/filters/FilterOperations.h:
(WebCore::FilterOperations::FilterOperations):
* Source/WebCore/rendering/PathOperation.cpp:
(WebCore::ReferencePathOperation::clone const):
(WebCore::RayPathOperation::clone const):
* Source/WebCore/rendering/PathOperation.h:
* Source/WebCore/rendering/style/BasicShapes.cpp:
(WebCore::BasicShapeCircle::clone const):
(WebCore::BasicShapeEllipse::clone const):
(WebCore::BasicShapePolygon::clone const):
(WebCore::BasicShapePath::clone const):
(WebCore::BasicShapeInset::clone const):
* Source/WebCore/rendering/style/BasicShapes.h:

Canonical link: https://commits.webkit.org/261418@main
